### PR TITLE
[IMP] mass_mailing(_sms): add calendar view & pimp schedule process

### PR DESCRIPTION
**Pull request by awa-odoo**
PURPOSE

To ease the scheduling process, we introduce a new calendar view on mailings
that allows to easily schedule your communications.
In addition, mailings/sms views were pimped to improve the global usability,
especially on the scheduling configuration.

SPECS

- add a calendar view to allow marketeers to either schedule or overview their
  ongoing mailings/sms ;
  Side note: inspired by what has been done with Social Posts
- improve the scheduling flow by allowing to configure it directly in the form
  view using fields rather than action buttons that open an extra window ;
- add a constraint on 'schedule_date' to make sure it's not scheduled in the
  past ;
- update mass_mailing_sms to match scheduling flow of emails and adjust the
  'sms_force_send' display to ease understanding ;

LINKS

Task ID 2202759

Co-authored-by: Aurélien Warnon <awa@odoo.com>

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/mass_mailing/models/mailing.py [hunk 1]
```diff
@@ -269,6 +306,11 @@ def create(self, values):
     def write(self, values):
         if values.get('body_html'):
             values['body_html'] = self._convert_inline_images_to_urls(values['body_html'])
+        if values.get('calendar_date'):
+            if any(mailing.schedule_type != 'scheduled' for mailing in self):
```
Comment (tde-banana-odoo) [inline addons/mass_mailing/models/mailing.py:310]: Couldn't we check that mailing is "in_queue" (and draft probably) state(s) instead ? By the way, how are displayed in calendar view unscheduled mailings (just put in queue) ?

For example we have a "next_departure" field. Maybe its logic could be integrated / copied into the schedule flow, order to always have a date: sent date, or scheduled, or "next departure". Then when moving mailing in calendar view, we force schedule_date, only if mailing is in draft or in queue.

That way we keep a state-based model ... not sure, just proposing.

**Review by tde-banana-odoo — COMMENTED** [review]
Some thoughts.

**Review by awa-odoo — COMMENTED** [review]
Code was modified according to the tips and every comment should be answered.
Don't hesitate to ask if something isn't clear, some concepts/flows are a bit hard to explain by text.

Thanks for the riviou!

File: addons/mass_mailing/models/mailing.py [hunk 2]
```diff
@@ -49,6 +49,14 @@ class MassMailing(models.Model):
     @api.model
     def default_get(self, fields):
         vals = super(MassMailing, self).default_get(fields)
+
+        default_calendar_date = self.env.context.get('default_calendar_date')
+        if default_calendar_date and ('schedule_type' in fields and 'schedule_date' in fields):
```
[the remark this answers is not in the document]
Comment (tde-banana-odoo) [reply addons/mass_mailing/models/mailing.py:54]: Ok. A small comment in code would be welcomed :D

File: addons/mass_mailing/views/mailing_mailing_views.xml [hunk 3]
```diff
@@ -56,9 +56,11 @@
             <field name="arch" type="xml">
                 <form string="Mailing">
                     <header style="min-height:31px;">
-                        <button name="action_put_in_queue" type="object" attrs="{'invisible': [('state', 'in', ('in_queue', 'done'))]}" class="oe_highlight" string="Send"
+                        <button id="action_launch_now" name="action_put_in_queue" type="object" class="oe_highlight" string="Launch"
```
Comment (tde-banana-odoo) [inline addons/mass_mailing/views/mailing_mailing_views.xml:59]: id="action_launch_now" name="action_put_in_queue"

Mmmh, launch now or put in queue ? What is the purpose of the id ?

With action_launch_scheduled button herebelow, seems like we could really improve a bit the naming / IDs of all that brol.

File: addons/mass_mailing/views/mailing_mailing_views.xml [hunk 4]
```diff
@@ -405,6 +418,18 @@
             </field>
         </record>
 
+        <record id="view_mail_mass_mailing_calendar" model="ir.ui.view">
```
Comment (tde-banana-odoo) [inline addons/mass_mailing/views/mailing_mailing_views.xml:421]: Tss tss, where are your guidelines ?

**Review by tde-banana-odoo — COMMENTED** [review]
Light comments !

**Comment by tde-banana-odoo** [comment]
@robodoo r+

**Pull request closed without merging**